In [8]:
import time
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
import pickle

import warnings
warnings.filterwarnings("ignore")
from sklearn.metrics import log_loss

In [3]:
train_features= pd.read_csv('train_features.csv')
print(train_features.shape)
train_features.head()

(23814, 876)


sig_id cp_type  cp_time cp_dose     g-0     g-1     g-2     g-3  \
0  id_000644bb2  trt_cp       24      D1  1.0620  0.5577 -0.2479 -0.6208   
1  id_000779bfc  trt_cp       72      D1  0.0743  0.4087  0.2991  0.0604   
2  id_000a6266a  trt_cp       48      D1  0.6280  0.5817  1.5540 -0.0764   
3  id_0015fd391  trt_cp       48      D1 -0.5138 -0.2491 -0.2656  0.5288   
4  id_001626bd3  trt_cp       72      D2 -0.3254 -0.4009  0.9700  0.6919   

      g-4     g-5  ...    c-90    c-91    c-92    c-93    c-94    c-95  \
0 -0.1944 -1.0120  ...  0.2862  0.2584  0.8076  0.5523 -0.1912  0.6584   
1  1.0190  0.5207  ... -0.4265  0.7543  0.4708  0.0230  0.2957  0.4899   
2 -0.0323  1.2390  ... -0.7250 -0.6297  0.6103  0.0223 -1.3240 -0.3174   
3  4.0620 -0.8095  ... -2.0990 -0.6441 -5.6300 -1.3780 -0.8632 -1.2880   
4  1.4180 -0.8244  ...  0.0042  0.0048  0.6670  1.0690  0.5523 -0.3031   

     c-96    c-97    c-98    c-99  
0 -0.3981  0.2139  0.3801  0.4176  
1  0.1522  0.1241  0.6077  0.7371  
2 -0.6417 -0.2187 -1.4080  0.6931  
3 -1.6210 -0.8784 -0.3876 -0.8154  
4  0.1094  0.2885 -0.3786  0.7125  

[5 rows x 876 columns]

In [4]:
test_features= pd.read_csv('test_features.csv')
print(test_features.shape)
test_features.head()

(3982, 876)


sig_id      cp_type  cp_time cp_dose     g-0     g-1     g-2     g-3  \
0  id_0004d9e33       trt_cp       24      D1 -0.5458  0.1306 -0.5135  0.4408   
1  id_001897cda       trt_cp       72      D1 -0.1829  0.2320  1.2080 -0.4522   
2  id_002429b5b  ctl_vehicle       24      D1  0.1852 -0.1404 -0.3911  0.1310   
3  id_00276f245       trt_cp       24      D2  0.4828  0.1955  0.3825  0.4244   
4  id_0027f1083       trt_cp       48      D1 -0.3979 -1.2680  1.9130  0.2057   

      g-4     g-5  ...    c-90    c-91    c-92    c-93    c-94    c-95  \
0  1.5500 -0.1644  ...  0.0981  0.7978 -0.1430 -0.2067 -0.2303 -0.1193   
1 -0.3652 -0.3319  ... -0.1190 -0.1852 -1.0310 -1.3670 -0.3690 -0.5382   
2 -1.4380  0.2455  ... -0.2261  0.3370 -1.3840  0.8604 -1.9530 -1.0140   
3 -0.5855 -1.2020  ...  0.1260  0.1570 -0.1784 -1.1200 -0.4325 -0.9005   
4 -0.5864 -0.0166  ...  0.4965  0.7578 -0.1580  1.0510  0.5742  1.0900   

     c-96    c-97    c-98    c-99  
0  0.0210 -0.0502  0.1510 -0.7750  
1  0.0359 -0.4764 -1.3810 -0.7300  
2  0.8662  1.0160  0.4924 -0.1942  
3  0.8131 -0.1305  0.5645 -0.5809  
4 -0.2962 -0.5313  0.9931  1.8380  

[5 rows x 876 columns]

In [5]:
train_drug= pd.read_csv('train_drug.csv')
print(train_drug.shape)
train_drug.head()

(23814, 2)


sig_id    drug_id
0  id_000644bb2  b68db1d53
1  id_000779bfc  df89a8e5a
2  id_000a6266a  18bb41b2c
3  id_0015fd391  8c7f86626
4  id_001626bd3  7cbed3131

In [6]:
train_targets_nonscored= pd.read_csv('train_targets_nonscored.csv')
print(train_targets_nonscored.shape)
train_targets_nonscored.head()

(23814, 403)


sig_id  abc_transporter_expression_enhancer  abl_inhibitor  \
0  id_000644bb2                                    0              0   
1  id_000779bfc                                    0              0   
2  id_000a6266a                                    0              0   
3  id_0015fd391                                    0              0   
4  id_001626bd3                                    0              0   

   ace_inhibitor  acetylcholine_release_enhancer  \
0              0                               0   
1              0                               0   
2              0                               0   
3              0                               0   
4              0                               0   

   adenosine_deaminase_inhibitor  adenosine_kinase_inhibitor  \
0                              0                           0   
1                              0                           0   
2                              0                           0   
3                              0                           0   
4                              0                           0   

   adenylyl_cyclase_inhibitor  age_inhibitor  alcohol_dehydrogenase_inhibitor  \
0                           0              0                                0   
1                           0              0                                0   
2                           0              0                                0   
3                           0              0                                0   
4                           0              0                                0   

   ...  ve-cadherin_antagonist  vesicular_monoamine_transporter_inhibitor  \
0  ...                       0                                          0   
1  ...                       0                                          0   
2  ...                       0                                          0   
3  ...                       0                                          0   
4  ...                       0                                          0   

   vitamin_k_antagonist  voltage-gated_calcium_channel_ligand  \
0                     0                                     0   
1                     0                                     0   
2                     0                                     0   
3                     0                                     0   
4                     0                                     0   

   voltage-gated_potassium_channel_activator  \
0                                          0   
1                                          0   
2                                          0   
3                                          0   
4                                          0   

   voltage-gated_sodium_channel_blocker  wdr5_mll_interaction_inhibitor  \
0                                     0                               0   
1                                     0                               0   
2                                     0                               0   
3                                     0                               0   
4                                     0                               0   

   wnt_agonist  xanthine_oxidase_inhibitor  xiap_inhibitor  
0            0                           0               0  
1            0                           0               0  
2            0                           0               0  
3            0                           0               0  
4            0                           0               0  

[5 rows x 403 columns]

In [7]:
train_targets_scored= pd.read_csv('train_targets_scored.csv')
print(train_targets_scored.shape)
train_targets_scored.head()

(23814, 207)


sig_id  5-alpha_reductase_inhibitor  11-beta-hsd1_inhibitor  \
0  id_000644bb2                            0                       0   
1  id_000779bfc                            0                       0   
2  id_000a6266a                            0                       0   
3  id_0015fd391                            0                       0   
4  id_001626bd3                            0                       0   

   acat_inhibitor  acetylcholine_receptor_agonist  \
0               0                               0   
1               0                               0   
2               0                               0   
3               0                               0   
4               0                               0   

   acetylcholine_receptor_antagonist  acetylcholinesterase_inhibitor  \
0                                  0                               0   
1                                  0                               0   
2                                  0                               0   
3                                  0                               0   
4                                  0                               0   

   adenosine_receptor_agonist  adenosine_receptor_antagonist  \
0                           0                              0   
1                           0                              0   
2                           0                              0   
3                           0                              0   
4                           0                              0   

   adenylyl_cyclase_activator  ...  tropomyosin_receptor_kinase_inhibitor  \
0                           0  ...                                      0   
1                           0  ...                                      0   
2                           0  ...                                      0   
3                           0  ...                                      0   
4                           0  ...                                      0   

   trpv_agonist  trpv_antagonist  tubulin_inhibitor  \
0             0                0                  0   
1             0                0                  0   
2             0                0                  0   
3             0                0                  0   
4             0                0                  0   

   tyrosine_kinase_inhibitor  ubiquitin_specific_protease_inhibitor  \
0                          0                                      0   
1                          0                                      0   
2                          0                                      0   
3                          0                                      0   
4                          0                                      0   

   vegfr_inhibitor  vitamin_b  vitamin_d_receptor_agonist  wnt_inhibitor  
0                0          0                           0              0  
1                0          0                           0              0  
2                0          0                           0              0  
3                0          0                           0              0  
4                0          0                           0              0  

[5 rows x 207 columns]

# FINAL PIPELINE WITH  BEST MODEL

In [8]:
train_features.head(2)

sig_id cp_type  cp_time cp_dose     g-0     g-1     g-2     g-3  \
0  id_000644bb2  trt_cp       24      D1  1.0620  0.5577 -0.2479 -0.6208   
1  id_000779bfc  trt_cp       72      D1  0.0743  0.4087  0.2991  0.0604   

      g-4     g-5  ...    c-90    c-91    c-92    c-93    c-94    c-95  \
0 -0.1944 -1.0120  ...  0.2862  0.2584  0.8076  0.5523 -0.1912  0.6584   
1  1.0190  0.5207  ... -0.4265  0.7543  0.4708  0.0230  0.2957  0.4899   

     c-96    c-97    c-98    c-99  
0 -0.3981  0.2139  0.3801  0.4176  
1  0.1522  0.1241  0.6077  0.7371  

[2 rows x 876 columns]

In [9]:
train_targets_scored.head(2)

sig_id  5-alpha_reductase_inhibitor  11-beta-hsd1_inhibitor  \
0  id_000644bb2                            0                       0   
1  id_000779bfc                            0                       0   

   acat_inhibitor  acetylcholine_receptor_agonist  \
0               0                               0   
1               0                               0   

   acetylcholine_receptor_antagonist  acetylcholinesterase_inhibitor  \
0                                  0                               0   
1                                  0                               0   

   adenosine_receptor_agonist  adenosine_receptor_antagonist  \
0                           0                              0   
1                           0                              0   

   adenylyl_cyclase_activator  ...  tropomyosin_receptor_kinase_inhibitor  \
0                           0  ...                                      0   
1                           0  ...                                      0   

   trpv_agonist  trpv_antagonist  tubulin_inhibitor  \
0             0                0                  0   
1             0                0                  0   

   tyrosine_kinase_inhibitor  ubiquitin_specific_protease_inhibitor  \
0                          0                                      0   
1                          0                                      0   

   vegfr_inhibitor  vitamin_b  vitamin_d_receptor_agonist  wnt_inhibitor  
0                0          0                           0              0  
1                0          0                           0              0  

[2 rows x 207 columns]

In [10]:
test_features.head(2)

sig_id cp_type  cp_time cp_dose     g-0     g-1     g-2     g-3  \
0  id_0004d9e33  trt_cp       24      D1 -0.5458  0.1306 -0.5135  0.4408   
1  id_001897cda  trt_cp       72      D1 -0.1829  0.2320  1.2080 -0.4522   

      g-4     g-5  ...    c-90    c-91   c-92    c-93    c-94    c-95    c-96  \
0  1.5500 -0.1644  ...  0.0981  0.7978 -0.143 -0.2067 -0.2303 -0.1193  0.0210   
1 -0.3652 -0.3319  ... -0.1190 -0.1852 -1.031 -1.3670 -0.3690 -0.5382  0.0359   

     c-97   c-98   c-99  
0 -0.0502  0.151 -0.775  
1 -0.4764 -1.381 -0.730  

[2 rows x 876 columns]

In [11]:
X = train_features.iloc[ :, 1:]
y = train_targets_scored.iloc[: , 1:]

In [12]:
X.head(2)

cp_type  cp_time cp_dose     g-0     g-1     g-2     g-3     g-4     g-5  \
0  trt_cp       24      D1  1.0620  0.5577 -0.2479 -0.6208 -0.1944 -1.0120   
1  trt_cp       72      D1  0.0743  0.4087  0.2991  0.0604  1.0190  0.5207   

      g-6  ...    c-90    c-91    c-92    c-93    c-94    c-95    c-96  \
0 -1.0220  ...  0.2862  0.2584  0.8076  0.5523 -0.1912  0.6584 -0.3981   
1  0.2341  ... -0.4265  0.7543  0.4708  0.0230  0.2957  0.4899  0.1522   

     c-97    c-98    c-99  
0  0.2139  0.3801  0.4176  
1  0.1241  0.6077  0.7371  

[2 rows x 875 columns]

In [13]:
y.head(2)

5-alpha_reductase_inhibitor  11-beta-hsd1_inhibitor  acat_inhibitor  \
0                            0                       0               0   
1                            0                       0               0   

   acetylcholine_receptor_agonist  acetylcholine_receptor_antagonist  \
0                               0                                  0   
1                               0                                  0   

   acetylcholinesterase_inhibitor  adenosine_receptor_agonist  \
0                               0                           0   
1                               0                           0   

   adenosine_receptor_antagonist  adenylyl_cyclase_activator  \
0                              0                           0   
1                              0                           0   

   adrenergic_receptor_agonist  ...  tropomyosin_receptor_kinase_inhibitor  \
0                            0  ...                                      0   
1                            0  ...                                      0   

   trpv_agonist  trpv_antagonist  tubulin_inhibitor  \
0             0                0                  0   
1             0                0                  0   

   tyrosine_kinase_inhibitor  ubiquitin_specific_protease_inhibitor  \
0                          0                                      0   
1                          0                                      0   

   vegfr_inhibitor  vitamin_b  vitamin_d_receptor_agonist  wnt_inhibitor  
0                0          0                           0              0  
1                0          0                           0              0  

[2 rows x 206 columns]

In [14]:
X_train, X_test, y_train, y_test = train_test_split(X , y,test_size=0.20)
X_train, X_cv, y_train, y_cv = train_test_split(X_train, y_train ,test_size=0.20)

In [15]:
print(X_train.shape , X_test.shape  , y_train.shape , y_test.shape)
print(X_train.shape , X_cv.shape  , y_train.shape , y_cv.shape)

(15240, 875) (4763, 875) (15240, 206) (4763, 206)
(15240, 875) (3811, 875) (15240, 206) (3811, 206)


In [16]:
print('Number of data points in train data:', X_train.shape[0])
print('Number of data points in test data:', X_test.shape[0])
print('Number of data points in cross validation data:', X_cv.shape[0])

Number of data points in train data: 15240
Number of data points in test data: 4763
Number of data points in cross validation data: 3811


In [5]:
#FUNCTION TO PREPROCESS DATA : X_TRAIN
def preprocess(train_features_new):
    #ENCODING THE CATEGORICAL DATA
    train_features_new['cp_type'] = train_features_new['cp_type'].map({'trt_cp' : 0 , 'ctl_vehicle' : 1})
    train_features_new['cp_time'] = train_features_new['cp_time'].map({24 : 0 , 72 : 1 , 48 : 2})
    train_features_new['cp_dose'] = train_features_new['cp_dose'].map({'D1' : 0 , 'D2' : 1})
    
    return train_features_new

In [6]:
#FUNCTION TO GET THE PREDICTED OUTPUT
def final(data_to_pred):

    start_time = time.time()
    pipe = pickle.load(open("MODLE_PIPE.pkl", 'rb')) #LOADING MY PRETRAINED MODEL
    data_processed = preprocess(data_to_pred)
    pred = pipe.predict(data_processed)#INPUT IS A DATAFRAME
    end_time = time.time()
    print('TIME TAKEN TO PREDICT IS : {}'.format(end_time - start_time))
    return pred

In [19]:
#DATA ISN'T PREPROCESSED , PRECEPROCESSING DONE WITHIN FUNCTION
pred_cv = final(X_cv) 
pred_test = final(X_test)

TIME TAKEN TO PREDICT IS : 236.9568154811859
TIME TAKEN TO PREDICT IS : 311.75401520729065


In [23]:
def metric(original , predicted):
    #GETTING THE METRIC
    print("LOGLOSS IS :",log_loss(original, predicted.toarray()))

In [24]:
metric(y_cv, pred_cv)
metric(y_test, pred_test)

LOGLOSS IS : 3.464779364335443
LOGLOSS IS : 3.5761113944689913


OBSERVATION :

WE GET A LOG LOSS OF  3.5761113944689913 WITH OUR BEST MODEL (MLKNN(K = 10)) FOR OUR X_TEST DATASET.

In [15]:
#THIS IS THE DATA FILE WHICH I LD GIVE AS INPUT IN MY BROWSER
data  = pd.read_csv('data')
print(data.cp_time.unique())
data

[24 72 48]


cp_type  cp_time cp_dose     g-0     g-1     g-2     g-3     g-4     g-5  \
0  trt_cp       24      D1  1.0620  0.5577 -0.2479 -0.6208 -0.1944 -1.0120   
1  trt_cp       72      D1  0.0743  0.4087  0.2991  0.0604  1.0190  0.5207   
2  trt_cp       48      D1  0.6280  0.5817  1.5540 -0.0764 -0.0323  1.2390   
3  trt_cp       48      D1 -0.5138 -0.2491 -0.2656  0.5288  4.0620 -0.8095   
4  trt_cp       72      D2 -0.3254 -0.4009  0.9700  0.6919  1.4180 -0.8244   
5  trt_cp       24      D1 -0.6111  0.2941 -0.9901  0.2277  1.2810  0.5203   
6  trt_cp       24      D2  2.0440  1.7000 -1.5390  5.9440 -2.1670 -4.0360   
7  trt_cp       48      D1  0.2711  0.5133 -0.1327  2.5950  0.6980  0.5846   
8  trt_cp       48      D1 -0.3014  0.5545 -0.2576 -0.1390 -0.6487 -0.6057   
9  trt_cp       48      D2 -0.0630  0.2564 -0.5279 -0.2541 -0.0182 -1.5370   

      g-6  ...    c-90    c-91    c-92    c-93    c-94    c-95    c-96  \
0 -1.0220  ...  0.2862  0.2584  0.8076  0.5523 -0.1912  0.6584 -0.3981   
1  0.2341  ... -0.4265  0.7543  0.4708  0.0230  0.2957  0.4899  0.1522   
2  0.1715  ... -0.7250 -0.6297  0.6103  0.0223 -1.3240 -0.3174 -0.6417   
3 -1.9590  ... -2.0990 -0.6441 -5.6300 -1.3780 -0.8632 -1.2880 -1.6210   
4 -0.2800  ...  0.0042  0.0048  0.6670  1.0690  0.5523 -0.3031  0.1094   
5  0.0543  ...  1.8390  1.1570 -1.0120  1.9010  1.4270  0.4519  1.2120   
6  3.6950  ...  0.1855  1.1720  0.8325  0.6486  0.8090  1.5880  2.4670   
7 -0.2633  ...  0.3230 -0.4140  0.2532  0.0513  0.8600  1.4250  0.6633   
8 -0.7549  ... -1.6470  0.2863  1.1070 -0.7735 -1.0280 -1.3070 -0.1167   
9 -0.2180  ...  0.2201  0.5601 -0.3501 -1.4070 -0.1717 -1.1160 -0.8745   

     c-97    c-98    c-99  
0  0.2139  0.3801  0.4176  
1  0.1241  0.6077  0.7371  
2 -0.2187 -1.4080  0.6931  
3 -0.8784 -0.3876 -0.8154  
4  0.2885 -0.3786  0.7125  
5  0.3765  0.7848  1.3990  
6  0.0357  0.1351 -0.3179  
7  0.4562 -0.9622  0.0260  
8 -0.1241 -0.6420  0.5543  
9 -0.2716  0.0189 -2.0000  

[10 rows x 875 columns]

In [16]:
start_time = time.time()
pipe = pickle.load(open("MODLE_PIPE.pkl", 'rb')) #LOADING MY PRETRAINED MODEL
data_processed = preprocess(data)
pred = pipe.predict(data_processed)#INPUT IS A DATAFRAME
print(pred.toarray())
print(type(pred))
print('done prediction')
end_time = time.time()
print('TIME TAKEN TO PREDICT IS : {}'.format(end_time - start_time))

[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]
<class 'scipy.sparse.lil.lil_matrix'>
done prediction
TIME TAKEN TO PREDICT IS : 0.8754982948303223
